fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/TemplateDruidQuery.java

```diff
@@ -429,6 +439,178 @@ public MetricField getMetricField(String name) {
                 .orElseThrow(IllegalArgumentException::new);
     }
 
+    /**
+     * Renames the {@link MetricField} with name {@code currentName} to {@code newName}, as well as any other
+     * MetricFields in this TemplateDruidQuery that reference it. This functionality is primarily meant to service query
+     * time metric renaming. As such, this method only renames metrics on the outermost TDQ. More involved query
+     * rewriting should be handled by custom {@link com.yahoo.bard.webservice.web.handlers.DataRequestHandler}
+     * implementations.
+     * <p>
+     * Renaming a metric to the same name (i.e. {@code currentName=foo, newName=foo}) is treated as a noop and this
+     * TemplateDruidQuery is returned with no operations performed on it.
+     *
+     * @param currentName  The name of the MetricField to be rewritten. This parameter cannot be null and a
+     *                     MetricField with this name must exist on this TemplateDruidQuery
+     * @param newName  The name for the target MetricField to be renamed to. This parameter cannot be null and cannot
+     *                 conflict with a MetricField already on this TemplateDruidQuery
+     * @return the TemplateDruidQuery with the target MetricField renamed with {@code newName}. If newName is equivalent
+     *         to currentName, this metric call is treated as a noop and this TemplateDruidQuery is returned
+     * @throws NullPointerException if either parameter is null
+     * @throws IllegalArgumentException if there is no MetricField that matches currentName, or if a MetricField that
+     *                                  matches newName already exists.
+     */
+    public TemplateDruidQuery renameMetricField(@NotNull String currentName, @NotNull String newName) {
+        Objects.requireNonNull(currentName, NULL_RENAME_ERROR_MESSAGE);
+        Objects.requireNonNull(newName, NULL_RENAME_ERROR_MESSAGE);
+
+        if (currentName.equals(newName)) {
+            return this;
+        }
+
+        if (Stream.concat(getAggregations().stream(), getPostAggregations().stream())
+                .anyMatch(mf -> Objects.equals(newName, mf.getName()))) {
+            throw new IllegalArgumentException(
+                String.format(RENAME_TO_DUPLICATE_NAME_ERROR_MESSAGE, currentName, newName)
+            );
+        }
+
+        if (getAggregations().stream().anyMatch(agg -> Objects.equals(agg.getName(), currentName))) {
+            return renameAggregation(currentName, newName);
+        }
+
+        if (getPostAggregations().stream().anyMatch(pa -> Objects.equals(pa.getName(), currentName))) {
+            return renamePostAggregation(currentName, newName);
+        }
+
+        throw new IllegalArgumentException(String.format(NO_METRIC_TO_RENAME_FOUND_ERROR_MESSAGE, currentName));
```

Comment:
I'm not sure how I feel about this. On the one hand, I think people would expect that it's safe to rename a
metric that doesn't exist on the current query, and may in fact try to do so because it means one less edge case
they have to worry about. On the other, this does allow us to catch stupid typos or other minor errors that would
otherwise pass unnoticed.

Reply:
Let's try being strict in the first pass, and if it's annoying we can add a warn-and-continue option.

Reply:
Let's add a gatekeeping boolean method to check for the existance of a field name so that someone who wants to be careful can do an if( exists(oldName) ) then rename(oldname, newname) pattern.

Reply:
Alternatively, add two methods, one which takes a flag for strict validation and the other that defaults strict to false.  But I think making this strict avoids unintentional errors well.

Reply:
I'm not sure if this feature is worth flagging. Might be worth reconsidering if we feature flag validation multiple different pieces of api request building.